cirq/testing/order_tester.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A utility class for testing ordering methods.
+
+To test an ordering method, create an OrderTester and add several
+equivalence groups or items to it. The order tester will check that
+the items within each group are all equal to each other, and every new
+added item or group is strictly ascending with regard to the previously
+added items or groups.
+
+It will also check that a==b implies hash(a)==hash(b).
+"""
+
+import collections
+
+from typing import Any
+
+import itertools
+
+
+class OrderTester:
+    """Tests ordering against user-provided disjoint ordered groups or items."""
+
+    def __init__(self):
+        self.groups = [(_ClassUnknownToSubjects(),)]
+
+    def _verify_ascending(self, v1, v2):
+        # __lt__ and __gt__ might not necessarily be implemented both.
```

Comment:
Check that the usual syntax works in all eight cases, which can each invoke different methods in different orders:

```python
assert v1 < v2
assert v1 <= v2
assert v2 > v1
assert v2 >= v1
assert not v1 > v2
assert not v2 >= v1
assert not v1 < v2
assert not v1 <= v2
```

Each of those methods may be implemented with different code (especially in python 2).

Then also check that all four of the `__lt__`/`__le__`/`__ge__`/`__gt__` methods works for both values (returns the correct thing or returns `NotImplemented` or is not present).

Reply:
Ok, so you are saying that we should assume that all relational operators are implemented.

Reply:
We should test that all relational binary operators work, because it is possible that they trigger different code. Testing the explicit methods is less important, and I don't think we should not require that they be present.

Reply:
Not-versions create problems for the initial add:
```        assert not hasattr(v1, '__le__') or not v1 <= v2```